substrate/frame/support/src/traits/hooks.rs

```diff
@@ -227,6 +227,57 @@ pub trait OnRuntimeUpgrade {
 	}
 }
 
+/// See [`Hooks::on_runtime_upgrade`].
+pub trait UncheckedOnRuntimeUpgrade {
+	/// See [`Hooks::on_runtime_upgrade`].
+	fn unchecked_on_runtime_upgrade() -> Weight {
+		Weight::zero()
+	}
+
+	/// See [`Hooks::pre_upgrade`].
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
+		Ok(Vec::new())
+	}
+
+	/// See [`Hooks::post_upgrade`].
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(_state: Vec<u8>) -> Result<(), TryRuntimeError> {
+		Ok(())
+	}
+
+	/// The expected and default behavior of this method is to handle executing `pre_upgrade` ->
+	/// `on_runtime_upgrade` -> `post_upgrade` hooks for a migration.
+	///
+	/// Internally, the default implementation
+	/// - Handles passing data from `pre_upgrade` to `post_upgrade`
+	/// - Ensure storage is not modified in `pre_upgrade` and `post_upgrade` hooks.
+	///
+	/// Combining the `pre_upgrade` -> `on_runtime_upgrade` -> `post_upgrade` logic flow into a
+	/// single method call is helpful for scenarios like testing a tuple of migrations, where the
+	/// tuple contains order-dependent migrations.
+	#[cfg(feature = "try-runtime")]
+	fn try_on_runtime_upgrade(checks: bool) -> Result<Weight, TryRuntimeError> {
```

Comment:
is this really needed? I think we shouldn't invoke try runtime on `UncheckedOnRuntimeUpgrade` directly. instead, they should be called on the checked one

Reply:
yeah, I couldn't find a way to not duplicate it.
if we want to pass a custom logic for `try-runtime` we need `UncheckedOnRuntimeUpgrade` to have these functions, since we can't control impl of `OnRuntimeUpgrade` in `VersionedMigration`. I was thinking of maybe passing a new type with `try-runtime` interface to `VersionedMigration`, i.e:

```rs
pub struct VersionedMigration<const FROM: u16, const TO: u16, Inner, TryRuntimeHooks, Pallet, Weight>

// inside `OnRuntimeUpgrade::pre_upgrade()` impl for `VersionedMigration`
#[cfg(feature = "try-runtime")]
fn pre_upgrade() -> Result<sp_std::vec::Vec<u8>, sp_runtime::TryRuntimeError> {
let on_chain_version = Pallet::on_chain_storage_version();
if on_chain_version == FROM {
	Ok(VersionedPostUpgradeData::MigrationExecuted(TryRuntimeHooks::pre_upgrade()?).encode())
} else {
	Ok(VersionedPostUpgradeData::Noop.encode())
}
}

```

Reply:
I don't really get it.
`UncheckedOnRuntimeUpgrade` should only be used by `VersionedMigration`
`VersionedMigration` already implements `OnRuntimeUpgrade`

I don't see any reason to have `try_on_runtime_upgrade` here

Reply:
ah yeah, got it. I thought you meant all functions under try-runtime. Will remove try-on-runtime-upgrade